Introduce event types to update/remove credentials

resah opened the pull request:
- Replace `UPDATE_TOTP` and `UPDATE_PASSWORD` with `UPDATE_CREDENTIAL` (and their respective error event types)
- Replace `REMOVE_TOTP` with `REMOVE_CREDENTIAL` (and its respective error event type)

Their event details contain the actual credential type.

All event type lists (e.g. in realm exports) have been adjusted accordingly.

Also documentation was adjusted when I could find it.

Closes #10114

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

resah committed:
introduce event types to update/remove credentials

Closes #10114

Signed-off-by: Theresa Henze <theresa.henze@bare.id>

ahus1 commented:
Hi @resah - thank you for this PR. Looking at the code changes, the renames look good to me, I'll still need to test them more thoroughly. 

Looking at the parent issue that this PR is about to close, it was about adding events for WebAuthn. I didn't see tests or code changes related to that, so I wonder: Is this addressed?

resah commented:
@ahus1

> Looking at the parent issue that this PR is about to close, it was about adding events for WebAuthn. I didn't see tests or code changes related to that, so I wonder: Is this addressed?

Thanks for reviewing this PR, I adjusted it by

- replacing the `CUSTOM_REQUIRED_ACTION` with `UPDATE_CREDENTIAL` in webauthn creation (and their tests). Since setting a password or a TOTP is also an `UPDATE_CREDENTIAL` event, this should be consistent. This way the email event listener will also be triggered for webauthn. 
- updating the email templates for the `UPDATE_CREDENTIAL` and `REMOVE_CREDENTIAL` events.

Please let me know, if you think the PR is missing something :)

Regards
Theresa

pedroigor commented on server-spi-private/src/main/java/org/keycloak/events/EventType.java:
```diff
@@ -62,16 +62,24 @@ public enum EventType implements EnumWithStableIndex {
     UPDATE_PROFILE(11, true),
     UPDATE_PROFILE_ERROR(0x10000 + UPDATE_PROFILE.getStableIndex(), true),
+
+    @Deprecated
```
We should still fire those events to allow people to migrate to the changes you are introducing here. We can consider removing them in Keycloak 27/28.

resah replied:
Hi @pedroigor , while I think, this is a good idea, I am unsure how to implement two events. 

Take for example [`src/main/java/org/keycloak/authentication/requiredactions/UpdateTotp.java`](https://github.com/keycloak/keycloak/blob/13826a0b63cff6bfd564a1ef2ed30a04304db259/services/src/main/java/org/keycloak/authentication/requiredactions/UpdateTotp.java#L75) which gets only a `RequiredActionContext` and the associated `EventBuilder`, but is **not** in charge of triggering the actual event (via `.success()`). The `EventBuilder` would overwrite the previously set event type and I don't see a possibility to provide two (or more) events this way, which are properly handled. Other code occurrences are quite similar to this.

Please let me know, if I am missing something, but imho this was probably a architectural design choice to only be able to have one event at the same time, which prevents a proper deprecation and migration phase for event types.

pedroigor replied:
You can fire a new event (or the existing one) after cloning the current one.

The main problem here is that we will likely break current deployments and the infrastructure they have around auditing. I'm also not sure if firing two events is a nice thing but at the same time I don't know how to not break things ...

resah replied:
@pedroigor Alright, it's definitely not pretty, but now there are two events, a new generic one and and the old one.

pedroigor reviewed (commented):
I also agree that having separate events for the different credentials does not make much sense if you can just infer the credential type from the event.

However, this is a breaking change and we need to communicate it properly in the release notes and upgrade/migration guide.

pedroigor commented:
@mposolda Wdyt about this proposal and have a single event for whatever credential we support?

mposolda commented:
@pedroigor @ahus1 @resah Sorry for late response.

I think that generic event like `UPDATE_CREDENTIAL` and `REMOVE_CREDENTIAL` is an ideal way to go. But I have few additional questions for the discussion:

- Do we want also dedicated event like `CREATE_CREDENTIAL` ? For example in case of "password", user is updating his current password, so the action like `UPDATE_CREDENTIAL` makes sense. Strictly speaking, for some other credentials like "otp" or "webauthn", user is rather creating new TOTP (or new WebAuthn key) in the account console rather than updating existing otp or WebAuthn. If he wants to update existing TOTP, he needs to create new TOTP and then remove the previous one. But not sure... Maybe using just `UPDATE_CREDENTIAL` is sufficient here as it is usually implicit whether it is "create" or "update" based on the credential type? What do others think?

- How to handle  backwards compatibility? I can see options like:

1. Make sure that updating password and OTP will create the duplicated events. So in addition to `UPDATE_CREDENTIAL`, there will be also `UPDATE_PASSWORD` or `UPDATE_TOTP`. Then the dedicated events like `UPDATE_PASSWORD`, `UPDATE_TOTP` and `REMOVE_TOTP` will be deprecated now and removed in some future release (Keycloak 27 or 28). 

2. Still keep using dedicated events like `UPDATE_PASSWORD`, `UPDATE_TOTP` and `REMOVE_TOTP` just for the password and TOTP credentials. For all other credential, use the `UPDATE_CREDENTIAL` (or `REMOVE_CREDENTIAL`).

My vote is (1), which is pretty much done in this PR if I understand correctly? But in that case, it may be good to also update "Migration guide" IMO and add the info that events like `UPDATE_PASSWORD`, `UPDATE_TOTP` and `REMOVE_TOTP` are deprecated.

resah commented:
@mposolda 

Thank you for your feedback!

> * Do we want also dedicated event like `CREATE_CREDENTIAL` ? For example in case of "password", user is updating his current password, so the action like `UPDATE_CREDENTIAL` makes sense. Strictly speaking, for some other credentials like "otp" or "webauthn", user is rather creating new TOTP (or new WebAuthn key) in the account console rather than updating existing otp or WebAuthn. If he wants to update existing TOTP, he needs to create new TOTP and then remove the previous one. But not sure... Maybe using just `UPDATE_CREDENTIAL` is sufficient here as it is usually implicit whether it is "create" or "update" based on the credential type? What do others think?

The actual creation or update of a credential in the specific implementations of the `CredentialProvider` know nothing about the current event and it would imho break a lot to propagate the actual process (create or update) back to the event. Additionally, for `otp` or `webauthn` there is only update and remove, but no create.

I think, the `UPDATE_CREDENTIAL` event is meant like notifying an update on the credential (type) "group" (e.g. "something changed in webauthn credentials") and not on a specific credential. What exactly was changed is then given as a credential id in the event details. 
Not sure, I can explain this right :)

> 1. Make sure that updating password and OTP will create the duplicated events. So in addition to `UPDATE_CREDENTIAL`, there will be also `UPDATE_PASSWORD` or `UPDATE_TOTP`. Then the dedicated events like `UPDATE_PASSWORD`, `UPDATE_TOTP` and `REMOVE_TOTP` will be deprecated now and removed in some future release (Keycloak 27 or 28).
> 
> My vote is (1), which is pretty much done in this PR if I understand correctly? But in that case, it may be good to also update "Migration guide" IMO and add the info that events like `UPDATE_PASSWORD`, `UPDATE_TOTP` and `REMOVE_TOTP` are deprecated.

Glad to hear, that you prefer (1), because this is what this merge request implements right now: The deprecated events are still triggered separately and marked as deprecated. We can remove those safely in a future version.
I also added a paragraph to the migration guide and the release notes about the new types and the deprecation of the old ones.

mposolda reviewed (dismissed):
@resah Thanks for the clarification.

@ahus1 @pedroigor I am approving the PR as the direction of this PR looks ok to me. WDYT? If we agree to merge this, then once it is merged, we can create follow-up issue for removing the deprecated events in Keycloak 27 .

ahus1 commented:
@resah - I see you are rebasing this PR regularly. Please don't do that unless there are conflicts in your PR. It would otherwise invalidate previous reviews, and it eats up resources on our CI pipeline. Thanks!

ssilvert reviewed (approved):
Approve UI parts.

ahus1 commented:
I'm happy to see this merged. Thank you for everyone involved in this!

mposolda commented:
For tracking: I've created https://github.com/keycloak/keycloak/issues/32685 for remove the deprecated password/OTP specific events in Keycloak 27
